File: src/modifiers/onGeometryChange.ts
import { createModifier } from './createModifier';

/// Size payload delivered to `onGeometryChange`'s handler.
export interface GeometrySize {
  width: number;
  height: number;
}

/// SwiftUI `.onGeometryChange(for:of:action:)` (watchOS 11+ / iOS 16+),
/// observing the view's `size`. The handler fires with the current size
/// whenever it changes.
///
/// ```tsx
/// <Rectangle modifiers={[onGeometryChange((s) => setSize(s))]} />
/// ```
///
/// On watchOS < 11 this is a no-op (view passes through, handler never
/// fires) — guarded natively.
export function onGeometryChange(handler: (size: GeometrySize) => void) {
  return createModifier('onGeometryChange', { handler });
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/OnGeometryChange.swift
import SwiftUI

/// SwiftUI `.onGeometryChange(for:of:action:)` (watchOS 11+ / iOS 16+),
/// observing the view's `size`. Fires the JS handler with
/// `{ width, height }` whenever the size changes. On watchOS < 11 this is a
/// no-op (view passes through unchanged, handler never fires).
enum RNWOnGeometryChangeModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("onGeometryChange") { view, params, ctx in
            let handlerId = params.int("handler")
            if #available(watchOS 11.0, *) {
                return AnyView(view.onGeometryChange(for: CGSize.self) { proxy in
                    proxy.size
                } action: { size in
                    ctx.fire(handlerId, [
                        "width": size.width,
                        "height": size.height,
                    ])
                })
            }
            return view
        }
    }
}
